### DISCUSSION TITLE
Catalog error:  Table with name d1_history does not exist!

### DISCUSSION THREAD

**Discussion by Megacannon** [duckdb/duckdb#12556] [Q&A]
I've ported my Python project over from Postgresql to DuckDB for the added speed.  I'm able to create tables and can write records to one of the tables, but for the life of me I can't figure out what's wrong with my d1_history table.  Here's the exact error:

`duckdb.duckdb.CatalogException: Catalog Error: Table with name d1_history does not exist!
Did you mean "pg_type"?`

d1_history absolutely does exist and I'm able to see it in DBeaver.  I'm also able to take the very query my code is trying to execute and run it successfully in DBeaver, so I know it's not something as simple as using unqualified names.

Things I've tried.
* using qualified names
* deleting and recreating the database
* Manually deleting the tables, recreating them in DBeaver instead of purely through SQL.
* Manually stepped through the code several times looking for anything suspicious.
* Explicitly disabled all Postgresql code I could find.
* Googled the crap out of it, but no one seems to have had the same error.  It just gives me a bunch of Postgresql stuff

For what it's worth, here's the table in question:

CREATE TABLE IF NOT EXISTS d1_history
(
    symbol character varying(10) NOT NULL,
    interval timestamp without time zone NOT NULL,
    open real NOT NULL,
    high real NOT NULL,
    low real NOT NULL,
    close real NOT NULL,
    adj_open real NOT NULL,
    adj_high real NOT NULL,
    adj_low real NOT NULL,
    adj_close real NOT NULL,
    volume bigint NOT NULL,
    split_coefficient real NOT NULL,
    CONSTRAINT d1_history_pkey PRIMARY KEY (symbol, interval)
)

I'd appreciate any help.  I've given up for the day.  I'm not sure what information you need, so just ask and I'll answer everything I can.

**Comment by grvhi** [comment]
Did you ever get to the bottom of this? I'm having exactly the same problem

**Comment by Tishj** [comment]
This got completely missed because it was raised as a discussion, not an issue

The code provided does not produce the issue you raise:
```sql
statement ok
CREATE TABLE IF NOT EXISTS d1_history
(
	symbol character varying(10) NOT NULL,
	interval timestamp without time zone NOT NULL,
	open real NOT NULL,
	high real NOT NULL,
	low real NOT NULL,
	close real NOT NULL,
	adj_open real NOT NULL,
	adj_high real NOT NULL,
	adj_low real NOT NULL,
	adj_close real NOT NULL,
	volume bigint NOT NULL,
	split_coefficient real NOT NULL,
	CONSTRAINT d1_history_pkey PRIMARY KEY (
		symbol,
		interval
	)
)
```

Please add additional information

**Comment by grvhi** [comment]
Hi @Tishj - I'm evidently not OP, but am seeing the same issue. I have attached some screenshots (from Pycharm) which show the Pycharm/Intellij Database viewer panel: that shows that the table exists in DuckDB as expected. However, using both SQLAlchemy (with the `duckdb-engine`) and `duckdb` directly, both report the same `CatalogException`, claiming the table does not exist. 

In my case, I'm experiencing this after adding a new table to an existing database (on disk).

I'm still experimenting, so will destroy the database and recreate with all tables (and reload data) to see if that results in the expected behaviour.  

![Screenshot 2024-10-14 at 8 35 10 AM](https://github.com/user-attachments/assets/4ad3bd9d-e3b7-43f2-9c80-72beb6fe86fb)
![Screenshot 2024-10-14 at 8 38 18 AM](https://github.com/user-attachments/assets/f060bc96-9aeb-4ba7-8d6a-a7a6f1ab399a)

**Reply by grvhi** [reply]
As a quick follow-up, I'm now seeing this issue for all tables, after destroying the database (deleting the file from disk) and recreating the tables. Which implies this is something environment specific. Perhaps related to concurrency (connecting to the database from Pycharm and the python process)? Although I do proactively disconnect Pycharm before interacting with the database from Python....

**Reply by grvhi** [reply]
Further, when I do `SHOW TABLES` via the pycharm database viewer, the command returns the expected tables. But I see the following in Python:

```python
>>> d = duckdb.connect('../db-name.duckdb')
>>> d.execute('SHOW TABLES;').fetchmany()
[]
```

This is the case with a brand new on-disk database. I have tried changing the filename of the persisted DB, but am seeing the same results. The `duckdb` package is version `1.1.2.dev87`

**Reply by grvhi** [reply]
If I switch to using an in-memory DB, I do not see any errors and all commands in the python environment return the expected results. So these errors seem to be limited to persisted databases.

**Reply by grvhi** [reply]
I see the same behaviour with `duckdb==1.1.1`

**Reply by grvhi** [reply]
Same behaviour with `duckdb==1.1.0`

**Reply by Tishj** [reply]
Thanks for the added info but none of this is reproducable

Can you submit the steps you use to create this fetch_history?

Also please dont provide images

**Reply by grvhi** [reply]
@Tishj - noted on the images. 

I'm using SQLAlchemy's ORM to create the tables, followed by the `create_all()` command. This always successfully creates the tables as expected, and I can browse the tables in the persisted DB with the DuckDB CLI and the Pycharm database viewer. Only the python environment reports that tables do not exist. 

```python
from sqlalchemy import (
    create_engine,
    String,
    ForeignKey,
    Integer,
    Sequence,
    Date, Float, UniqueConstraint, DateTime, Boolean
)
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    relationship
)

engine = create_engine('duckdb:///db-name.duckdb')

class FetchHistory(Base):
    __tablename__ = 'fetches'

    id: Mapped[int] = mapped_column(
        Integer,
        Sequence('execution_id_seq'),
        primary_key=True
    )
    date: Mapped[datetime.date] = mapped_column(Date())
    success: Mapped[bool] = mapped_column(Boolean(), default=False)
    total_accounts: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_changes: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_over_budget: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_no_budget: Mapped[int] = mapped_column(Integer(), nullable=True)


Base.metadata.create_all(engine)
```

**Reply by Tishj** [reply]
```py
import datetime

from sqlalchemy import (
    create_engine,
    String,
    ForeignKey,
    Integer,
    Sequence,
    Date, Float, UniqueConstraint, DateTime, Boolean
)
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    relationship,
    Session
)

engine = create_engine('duckdb:///db-name.duckdb')

class Base(DeclarativeBase):
    pass

class FetchHistory(Base):
    __tablename__ = 'fetches'

    id: Mapped[int] = mapped_column(
        Integer,
        Sequence('execution_id_seq'),
        primary_key=True
    )
    date: Mapped[datetime.date] = mapped_column(Date())
    success: Mapped[bool] = mapped_column(Boolean(), default=False)
    total_accounts: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_changes: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_over_budget: Mapped[int] = mapped_column(Integer(), nullable=True)
    total_no_budget: Mapped[int] = mapped_column(Integer(), nullable=True)


Base.metadata.create_all(engine)

with Session(bind=engine) as session:
	query = session.query(FetchHistory)
	query = query.order_by(FetchHistory.date.desc())
	print(query)
	latest_fetch = query.first()
	session.close()
print(latest_fetch)
```

does not error for me

**Reply by grvhi** [reply]
@Tishj - thank you for working through this with me. The problem actually exists with how I implemented `create_engine` with `duckdb-engine`. In short: relative paths in `create_engine` do not work as I had expected (i.e. the path becomes relative to the path of the file/module being executed, not the file which contains `create_engine`). On reflection, this makes sense, of course. Sorry for the run-around and thank you for taking the time to investigate.
